woocommerce/woocommerce
# [REST API] Improve product stock handling when managing orders

### Changes proposed in this Pull Request:

This makes a change to how product stock is handled when working with order line items via the REST API. It ensures that the stock level is updated when a line item is added or updated, but only when the corresponding order has one of the specified statuses. This makes the REST API flow more closely match the WP Admin flow.

Fixes #51024
Fixes #51974

### How to test the changes in this Pull Request:

We need to test that the changes in this PR work correctly, and also that they don't introduce any regressions.

#### Setup

* Set up your credentials for making API requests. Either create an API key or an app password for your user.
* Create a product (you can do this step via WP Admin). Under the Inventory tab, check the box to track stock quantity, and put in 50 or some other quantity that can vary over many requests.
* Leave a browser tab open with the list of products so you can refresh it and check if the inventory quantity has changed after each test request.

#### Tests

* Create a new order with a line item for the test product. Use `on-hold` for the order status.
    `POST wc/v3/orders`
    ```
    {
      "status": "on-hold",
      "line_items": [
        {
          "product_id": <product ID>,
          "quantity": 1
        }
      ]
    }
    ```
    _The product stock quantity should be reduced._
* Create a new order without a line item, but with the same `on-hold` status. Then in a separate request, update the new order with the line item.
    `POST wc/v3/orders/<order ID>`
    _The quantity should be reduced after making the update request._
* Update the order again and change the line item quantity to `2`. Make sure you include the line item ID, so it updates that line item instead of adding a new one.
    ```
        {
          "id": <line item ID>,
          "product_id": <product ID>,
          "quantity": 2
        }
    ```
    _The quantity should be reduced again._
* Update the order again and change the line item quantity to `0`.
    _The quantity should be increased back to its original amount. The request response should show that the line item has been removed from the order._
* Create a new order with a line item for the test product. This time set the order status as `auto-draft`.
    _The product stock quantity should **not** be reduced._
* Update the `auto-draft` order to set the line item quantity to `2`.
    _The quantity should still not be reduced._
* Update the order again to remove the line item.
    _The quantity should still not change._
* Update the order again to add the line item back in. Then in a separate request, update the order to change the order status to `processing`. Don't include the line items in the request payload, only the status.
    _The quantity should now be reduced by the amount in the line item._
* Go and view the orders created during this process in WP Admin and make sure the order notes accurately reflect all the stock quantity changes that were made.

**Are there any other cases or regressions we should be testing for here?**

Note that deleting the order altogether via the REST API (not just the line item) does not currently adjust the stock quantities. This is being handled in [a separate issue](https://github.com/woocommerce/woocommerce/issues/51002).

### Changelog entry












<details>

<summary>Changelog Entry Details</summary>

#### Significance




#### Type




#### Message 

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment 

</details>
